datasette/utils/asgi.py

```diff
@@ -0,0 +1,344 @@
+import json
+from datasette.utils import RequestParameters
+from mimetypes import guess_type
+from urllib.parse import parse_qs, urlunparse
+from pathlib import Path
+from html import escape
+import re
+import aiofiles
+
+
+class NotFound(Exception):
+    pass
+
+
+class Request:
+    def __init__(self, scope):
+        self.scope = scope
+
+    @property
+    def method(self):
+        return self.scope["method"]
+
+    @property
+    def url(self):
+        return urlunparse(
+            (self.scheme, self.host, self.path, None, self.query_string, None)
+        )
+
+    @property
+    def scheme(self):
+        return self.scope.get("scheme") or "http"
+
+    @property
+    def headers(self):
+        return dict(
+            [
+                (k.decode("latin-1").lower(), v.decode("latin-1"))
+                for k, v in self.scope.get("headers") or []
+            ]
+        )
+
+    @property
+    def host(self):
+        return self.headers.get("host") or "localhost"
+
+    @property
+    def path(self):
+        return (
+            self.scope.get("raw_path", self.scope["path"].encode("latin-1"))
+        ).decode("latin-1")
+
+    @property
+    def query_string(self):
+        return (self.scope.get("query_string") or b"").decode("latin-1")
+
+    @property
+    def args(self):
+        return RequestParameters(parse_qs(qs=self.query_string))
+
+    @property
+    def raw_args(self):
+        return {key: value[0] for key, value in self.args.items()}
+
+    @classmethod
+    def from_path_with_query_string(
```

Comment:
I'm sure I could come up with a snappier name for this method. How about one of these?

* `Request.mock("/path?foo=bar")`
* `Request.stub("/path?foo=bar")`
* `Request.fake("/path?foo=bar")`
* `Request.for("/path?foo=bar")`

Reply:
https://martinfowler.com/articles/mocksArentStubs.html suggests I should use `fake` here:

<img width="837" alt="Mocks_Aren_t_Stubs" src="https://user-images.githubusercontent.com/9599/59983619-cdcfd300-95d6-11e9-9717-8fe524977375.png">

Reply:
Renamed to `Request.fake(...)`